compiler/rustc_resolve/src/ident.rs

```diff
@@ -902,7 +908,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                 }
             }
 
-            self.record_use(ident, binding, restricted_shadowing);
+            self.record_use(ident, binding, used);
```

Comment:
This is the point where `if restricted_shadowind { Used::Scope } else { Used::Other }` wasn't precise enough?
That's why an extra `used: Option<Used>` parameter was added to this function?

Reply:
Yes, the main difference is here: 

```rust
Scope::Module(module, derive_fallback_lint_id) => {
                        let adjusted_parent_scope = &ParentScope { module, ..*parent_scope };
                        let binding = this.resolve_ident_in_module_unadjusted_ext(
                            ModuleOrUniformRoot::Module(module),
                            ident,
                            ns,
                            adjusted_parent_scope,
                            !matches!(scope_set, ScopeSet::Late(..)),
                            finalize,
                            ignore_binding,
                            Some(Used::Scope),
                        );

```

For Example:

```rust
use std::mem::drop;
fn main() {
    let s = Box::new(32);
    drop(s); 
}
```
 When resolve expr  ` drop(s);`, `restricted_shadowind` is !matches!(scope_set, ScopeSet::Late(..)), which is false, But used should be `Used::Scope`. Or rudundant import will not be reported.

The detailed log is as follows：


```bash
 0ms DEBUG rustc_resolve (recording res) recording PartialRes { base_res: Local(NodeId(17)), unresolved_segments: 0 } for 17
│ ├─  0ms DEBUG rustc_resolve (recording pat) recording NodeId(17) for ./src/main.rs:4:9: 4:10 (#0)
│ ├─┐rustc_resolve::late::smart_resolve_path_fragment qself=None, path=[Segment { ident: drop#0, id: Some(NodeId(26)), has_generic_args: false, has_lifetime_args: false, args_span: no-location (#0) }], source=Expr(Some(Expr { id: NodeId(24), kind: Call(Expr { id: NodeId(25), kind: Path(None, Path { span: ./src/main.rs:5:5: 5:9 (#0), segments: [PathSegment { ident: drop#0, id: NodeId(26), args: None }], tokens: None }), span: ./src/main.rs:5:5: 5:9 (#0), attrs: [], tokens: None }, [Expr { id: NodeId(27), kind: Path(None, Path { span: ./src/main.rs:5:10: 5:11 (#0), segments: [PathSegment { ident: s#0, id: NodeId(28), args: None }], tokens: None }), span: ./src/main.rs:5:10: 5:11 (#0), attrs: [], tokens: None }]), span: ./src/main.rs:5:5: 5:12 (#0), attrs: [], tokens: None })), finalize=Finalize { node_id: NodeId(25), path_span: ./src/main.rs:5:5: 5:9 (#0), root_span: ./src/main.rs:5:5: 5:9 (#0), report_private: true }, record_partial_res=Yes
│ │ ├─  0ms DEBUG rustc_resolve::late resolve_qpath(qself=None, path=[Segment { ident: drop#0, id: Some(NodeId(26)), has_generic_args: false, has_lifetime_args: false, args_span: no-location (#0) }], ns=ValueNS, finalize=Finalize { node_id: NodeId(25), path_span: ./src/main.rs:5:5: 5:9 (#0), root_span: ./src/main.rs:5:5: 5:9 (#0), report_private: true })
│ │ ├─  0ms DEBUG rustc_resolve::ident resolve_path ident 0 drop#0 Some(NodeId(26))
│ │ ├─┐rustc_resolve::ident::resolve_ident_in_lexical_scope ident=drop#0, ns=ValueNS, parent_scope=ParentScope { module: Some(Def(Mod, DefId(0:0 ~ main[c2f3]))), expansion: expn0, macro_rules: Cell { value: Empty }, derives: [] }, finalize=Some(Finalize { node_id: NodeId(25), path_span: ./src/main.rs:5:5: 5:9 (#0), root_span: ./src/main.rs:5:5: 5:9 (#0), report_private: true }), ignore_binding=None
│ │ │ ├─  0ms DEBUG rustc_resolve::ident walk rib
│ │ │ │ {s#0: Local(NodeId(17))}
│ │ │ ├─  0ms DEBUG rustc_resolve::ident walk rib
│ │ │ │ {}
│ │ │ ├─  0ms DEBUG rustc_resolve::ident walk rib
│ │ │ │ {}
│ │ │ ├─  0ms DEBUG rustc_resolve::ident walk rib
│ │ │ │ {}
│ │ │ ├─┐rustc_resolve::ident::early_resolve_ident_in_lexical_scope orig_ident=drop#0, scope_set=Late(ValueNS, Some(Def(Mod, DefId(0:0 ~ main[c2f3]))), Some(NodeId(25))), parent_scope=ParentScope { module: Some(Def(Mod, DefId(0:0 ~ main[c2f3]))), expansion: expn0, macro_rules: Cell { value: Empty }, derives: [] }, finalize=Some(Finalize { node_id: NodeId(25), path_span: ./src/main.rs:5:5: 5:9 (#0), root_span: ./src/main.rs:5:5: 5:9 (#0), report_private: true }), force=true, ignore_binding=None
│ │ │ │ ├─┐rustc_resolve::ident::resolve_ident_in_module_unadjusted_ext module=Module(Some(Def(Mod, DefId(0:0 ~ main[c2f3])))), ident=drop#0, ns=ValueNS, parent_scope=ParentScope { module: Some(Def(Mod, DefId(0:0 ~ main[c2f3]))), expansion: expn0, macro_rules: Cell { value: Empty }, derives: [] }, restricted_shadowing=false, finalize=Some(Finalize { node_id: NodeId(25), path_span: ./src/main.rs:5:5: 5:9 (#0), root_span: ./src/main.rs:5:5: 5:9 (#0), report_private: true }), ignore_binding=None, used=Some(Scope)
│ │ │ │ │ ├─  0ms DEBUG rustc_resolve::ident resolve_ident_in_module_unadjusted_ext ************************************************Please note 
 
 
 
 **************************************************************************************************************************************************************************************restricted_shadowing doesn't match Used  ident:drop#0 restricted_shadowing:false used:Some(Scope) binding:NameBindingData { kind: Import { binding: NameBindingData { kind: Res(Def(Fn, DefId(2:1641 ~ core[d3e4]::mem::drop))), ambiguity: None, warn_ambiguity: false, expansion: expn0, span: D:\source\rust\rust\library\core\src\mem\mod.rs:992:1: 992:22 (#0), vis: Public }, import: ImportData { kind: Single { source: drop#0, target: drop#0, source_bindings: PerNS { value_ns: Ok(..), type_ns: Err(Determined), macro_ns: Err(Determined) }, target_bindings: PerNS { value_ns: Some(..), type_ns: None, macro_ns: None }, type_ns_only: false, nested: false, id: NodeId(9) }, root_id: NodeId(9), use_span: ./src/main.rs:1:1: 1:20 (#0), use_span_with_attributes: ./src/main.rs:1:1: 1:20 (#0), has_attributes: false, span: ./src/main.rs:1:5: 1:19 (#0), root_span: ./src/main.rs:1:5: 1:19 (#0), parent_scope: ParentScope { module: Some(Def(Mod, DefId(0:0 ~ main[c2f3]))), expansion: expn0, macro_rules: Cell { value: Empty }, derives: [] }, module_path: [Segment { ident: {{root}}#0, id: None, has_generic_args: false, has_lifetime_args: false, args_span: no-location (#0) }, Segment { ident: std#0, id: Some(NodeId(10)), has_generic_args: false, has_lifetime_args: false, args_span: no-location (#0) }, Segment { ident: mem#0, id: Some(NodeId(11)), has_generic_args: false, has_lifetime_args: false, args_span: no-location (#0) }], imported_module: Cell { value: Some(Module(Some(Def(Mod, DefId(2:1481 ~ core[d3e4]::mem))))) }, vis: Cell { value: Some(Restricted(DefId(0:0 ~ main[c2f3]))) }, used: Cell { value: None } } }, ambiguity: None, warn_ambiguity: false, expansion: expn0, span: ./src/main.rs:1:5: 1:19 (#0), vis: Restricted(DefId(0:0 ~ main[c2f3])) }

```